### DISCUSSION TITLE
Using row number as an index, or creating expression column with row number

### DISCUSSION THREAD

**Discussion by sv2** [perspective-dev/perspective#2207] [Q&A] [answered]
Is there a way to use row number as an index in the table ? 
I am looking for a way to update specific row in an arbitrary dataset, where index column may or may not exist. 
To do so, I am adding a column with the row number - by exporting data via view.to_json(), then adding column with the row number to json, and then loading resulting json into the new table. 
Looking for a more efficient way to do this ... 
One idea  how to slightly optimize this - configure expression column with row number in the view, and then export to json and import to table, thus avoiding json manipulation. But I am not sure how to configure such expression column using ExprTk
Would appreciate any help and suggestions !

**Comment by texodus** [comment, marked as answer]
Internally perspective creates this column already. We should expose this column as a normal perspective column, but don't currently, so there's no way today to do this entirely in-engine today.

You can currently access the internal index column in un-grouped View's only, by providing the `id` flag when calling a `to_*()` method, where you'll find the row's internal id in the `__ID__` field:

```
const json = view.to_json({id: true})
const first_row_id = json[0].__ID__;
```

You can't yet update on it though, e.g. `table.update({__ID__: [12], ...})` won't work today, so you'll still need to mutate the data for now.

We're also adding `id()` as an exprtk function, but this is more of a hack for your use case wrt the former.

**Reply by sv2** [reply]
Many thanks !
